src/frontend/dash/circlePacker.ts

```diff
@@ -30,101 +38,149 @@ export class MemberCircle {
+        return;
     }
+    
+    for(let index1 = 0; index1 != placedCircles.length; index1++) {
+        const circle1 = placedCircles[index1];
+        for(let index2 = index1+1; index2 != placedCircles.length; index2++) {
+            const circle2 = placedCircles[index2];
+            const distanceFrom = circle1.r + circle2.r + MARGIN; 
+
+            const radius1 = circle1.r + circle.r + MARGIN;
+            const radius2 = circle2.r + circle.r + MARGIN;
+
+            // a : distance from fulcrum
+            const a = (radius1*radius1 - radius2*radius2 + distanceFrom*distanceFrom)/(2*distanceFrom);
+
+            // dx : delta X
+            const dx = circle1.x-circle2.x;
+            const dy = circle1.y-circle2.y;
+
+            // p : slope for perpendicular bisector of circle1 and circle2
+            const p = -dx/dy;
+
+            //
+            const multiplier = a/distanceFrom;
 
-    // recalc bounds
-    maxX = Math.max(maxX, circle.x + circle.r)
-    maxY = Math.max(maxY, circle.y + circle.r)
-    minX = Math.min(minX, circle.x - circle.r)
-    minY = Math.min(minY, circle.y - circle.r)
+            let posX = circle1.x - multiplier * dx;
+            let posY = circle1.y - multiplier * dy;
 
-    // redrawCircles(placedCircles.concat([circle])) // REDRAWWWWW = = = == = = = == = == = = = == 
-    // await sleep(10) /// REDRAWWWWW
+            let circleX = posX + Math.sqrt(radius1*radius1 - a*a)/Math.sqrt(1 + p*p);
+            let circleY = posY + p * Math.sqrt(radius1*radius1 - a*a)/Math.sqrt(1 + p*p);
+
+            if(isVacant(circle, circleX, circleY)) {
+                circle.x = circleX;
+                circle.y = circleY;
+                // circle1.touching[circle1.touching.length] = circle;
+                if(targetMaxX > circle.x + circle.r && targetMaxY > circle.y + circle.r  && targetMinX < circle.x - circle.r && targetMinY < circle.y - circle.r){
+                    return;
+                }
+                    
+            }
+
+            // console.log(a)
+            // console.log(p)
+            // console.log(radius1)
+            // console.log(radius2)
+
+            // improvements?
+            circleX = posX - Math.sqrt(radius1*radius1 - a*a)/Math.sqrt(1 + p*p);
+            circleY = posY - p * Math.sqrt(radius1*radius1 - a*a)/Math.sqrt(1 + p*p);
+
+            if(isVacant(circle, circleX, circleY)) {
+                circle.x = circleX;
+                circle.y = circleY;
+                if(targetMaxX > circle.x + circle.r && targetMaxY > circle.y + circle.r  && targetMinX < circle.x - circle.r && targetMinY < circle.y - circle.r) {
+                    return;
+                }
+            }
+        }
+    }
 }
 
+function isVacant(circle, x, y) {
+    for(let circle1 of placedCircles) {
+        // I'm afraid of inccuracies
+        if(getDistanceFrom(circle1, x, y) + 0.00001 < circle.r + circle1.r + MARGIN)
```

Comment:
Could you elaborate here on what the 0.00001 is doing?

Reply:
I forgot to take that out after I re added it when Elo gaslighted me while I was debugging his change

Reply:
![image](https://user-images.githubusercontent.com/78449263/200737997-b8b3c198-efdf-496c-b5ef-51bc02eca23d.png)